### PR TITLE
3D mortar mesh adjustments

### PR DESCRIPTION
**Pull request by lindsayad**
Depending on the memory layout a given process may store pointers in a different order from simulation run to run. By comparing against object id, we can be sure that the simulation will behave identically (with respect to how this particular data structure is sorted and in any code whose behavior is determined by the sorting of said structure) from run to run. Hence why we want to use sets with comparators that look at ids such that we can be sure of run-to-run reproducibility.

We are also choosing not to `prepare_for_use` because we only care about our own local numbering scheme and we do not care about partitioning (it's entirely local), neighbors, or unique ids

Refs #19092

### CODE REVIEW

**Comment by lindsayad** [comment]
With removal of `prepare_for_use`, we'll see if we can still write out the debug mesh...

**Comment by jbadger95** [comment]
Nice find, that makes sense. I'll put sets of pointers on my list of no-no's!

**Comment by recuero** [comment]
> I think that instead of succeeding every time, which this test now does, if we got "unlucky" and the xth processing of a primary element on processor 0 happened to create 5 new nodes, and the xth processing of a primary element on processor 1 happened to create 3 new nodes, then we would suddenly fail every time. 

@lindsayad I don't understand how removing `prepare_for_use` will get rid of the problem in general (but I am not familiar with that method).

**Comment by lindsayad** [comment]
`prepare_for_use` is the method that checks the sanity/consistency of `id` and `unique_id` numbering. But since @jbadger95 has essentially his own numbering scheme and because it doesn't actually impact simulation results whether we prepare or not (as he's said: a `Mesh` is really only a convenient data structure for holding the mortar segment nodes and elements), it's fine if we simply don't `prepare_for_use` (hopefully)

**Review by recuero — APPROVED** [review]

**Review by jbadger95 — APPROVED** [review]
Looks great!

**Review by dschwen — APPROVED** [review]
:+1: nice find and fix

File: framework/src/constraints/AutomaticMortarGeneration.C [hunk 1]
```diff
@@ -853,13 +853,24 @@ AutomaticMortarGeneration::buildMortarSegmentMesh3d()
...
+      struct CompareElemsByID
+      {
+        bool operator()(const Elem * a, const Elem * b) const
+        {
+          libmesh_assert(a);
+          libmesh_assert(b);
+
+          return a->id() < b->id();
+        }
+      };
```
[the remark this answers is not in the document]
Comment (dschwen) [reply framework/src/constraints/AutomaticMortarGeneration.C:839]: We have 89 occurrences of sets of pointers in framework and another 155 in modules. Most of those are probably benign, but at some point we should take a closer look. In modules most are XFEM element pointers and the rest are element and node pointers. All of those could impact order of operations leading to slightly non-deterministic results. I assume that we absorb those with our exodiff/csvdiff tolerances though.

[the remark this answers is not in the document]
Comment (lindsayad) [reply framework/src/constraints/AutomaticMortarGeneration.C:839]: Yea just using MPI leads to slightly non-deterministic results. It's only when there is code that relies on consistency that this is critical. And yea it's possible that there are some bad apples in those 244 occurrences but as you say I hope that most of them are benign

**Pull request merged**